File: python/src/trove_sdk/webhooks.py
"""
Verify Trove webhook signatures.

The Trove API signs every webhook delivery with HMAC-SHA256:

    X-Trove-Signature: t=<unix>,v1=<hex>

Where `v1 = hmac_sha256(secret, f"{t}.{raw_body}")`. To verify a delivery,
constant-time-compare the computed digest with `v1` and reject if `t` is too
old (default tolerance: 5 minutes).
"""
from __future__ import annotations

import hashlib
import hmac
import json
import time
from typing import Any, Union

from .exceptions import TroveError
from .models import WebhookEvent

_DEFAULT_TOLERANCE = 300  # seconds


class WebhookSignatureError(TroveError):
    """Raised when a webhook signature is malformed, mismatched, or stale."""


def _parse_header(signature_header: str) -> tuple[int, str]:
    """Pull `t=` and `v1=` out of `t=...,v1=...`. Order- and space-tolerant."""
    parts: dict[str, str] = {}
    for chunk in signature_header.split(","):
        chunk = chunk.strip()
        if "=" not in chunk:
            continue
        k, _, v = chunk.partition("=")
        parts[k.strip()] = v.strip()
    try:
        timestamp = int(parts["t"])
    except (KeyError, ValueError):
        raise WebhookSignatureError("Missing or invalid `t` in X-Trove-Signature")
    if "v1" not in parts:
        raise WebhookSignatureError("Missing `v1` in X-Trove-Signature")
    return timestamp, parts["v1"]


def verify_webhook(
    *,
    secret: str,
    body: Union[str, bytes],
    signature_header: str,
    tolerance_seconds: int = _DEFAULT_TOLERANCE,
) -> WebhookEvent:
    """Verify a Trove webhook delivery and return the parsed event.

    Raises `WebhookSignatureError` on bad signature, missing fields, or stale
    timestamp. Pass the *raw* request body — JSON re-serialization will
    invalidate the signature.

    Example (FastAPI):
        @app.post("/trove-webhook")
        async def receive(request: Request):
            body = await request.body()
            event = verify_webhook(
                secret=os.environ["TROVE_WEBHOOK_SECRET"],
                body=body,
                signature_header=request.headers["x-trove-signature"],
            )
            ...
    """
    if not signature_header:
        raise WebhookSignatureError("Missing X-Trove-Signature header")
    timestamp, provided = _parse_header(signature_header)

    age = abs(int(time.time()) - timestamp)
    if age > tolerance_seconds:
        raise WebhookSignatureError(
            f"Signature timestamp outside tolerance ({age}s > {tolerance_seconds}s)"
        )

    raw = body.encode() if isinstance(body, str) else body
    payload = f"{timestamp}.".encode() + raw
    expected = hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()
    if not hmac.compare_digest(expected, provided):
        raise WebhookSignatureError("Signature mismatch")

    try:
        decoded: dict[str, Any] = json.loads(raw)
    except json.JSONDecodeError as e:
        raise WebhookSignatureError(f"Body is not valid JSON: {e}")

    return WebhookEvent(
        id=decoded["id"],
        type=decoded["type"],
        api_version=decoded["api_version"],
        workspace_id=decoded["workspace_id"],
        namespace=decoded.get("namespace"),
        created_at=decoded["created_at"],
        data=decoded.get("data", {}),
    )

File: node/src/webhooks.ts
import { createHmac, timingSafeEqual } from 'node:crypto'
import { TroveError } from './error.js'
import type { WebhookEvent } from './types.js'

const DEFAULT_TOLERANCE_SECONDS = 300

export class WebhookSignatureError extends TroveError {
  constructor(message: string) {
    super(message, 400)
    this.name = 'WebhookSignatureError'
  }
}

function parseHeader(header: string): { timestamp: number; v1: string } {
  const parts: Record<string, string> = {}
  for (const chunk of header.split(',')) {
    const trimmed = chunk.trim()
    const eq = trimmed.indexOf('=')
    if (eq < 0) continue
    parts[trimmed.slice(0, eq).trim()] = trimmed.slice(eq + 1).trim()
  }
  const timestamp = parseInt(parts.t ?? '', 10)
  if (!Number.isFinite(timestamp)) {
    throw new WebhookSignatureError('Missing or invalid `t` in X-Trove-Signature')
  }
  if (!parts.v1) {
    throw new WebhookSignatureError('Missing `v1` in X-Trove-Signature')
  }
  return { timestamp, v1: parts.v1 }
}

export interface VerifyWebhookOptions {
  secret:           string
  body:             string | Uint8Array
  signatureHeader:  string
  toleranceSeconds?: number
}

/**
 * Verify a Trove webhook delivery and return the parsed event. Throws
 * `WebhookSignatureError` on bad signature, missing fields, or stale timestamp.
 *
 * Pass the *raw* request body — JSON re-serialization will invalidate the
 * signature.
 *
 * Example (Express):
 *     app.post('/trove-webhook', express.raw({type: 'application/json'}), (req, res) => {
 *       const event = verifyWebhook({
 *         secret: process.env.TROVE_WEBHOOK_SECRET!,
 *         body: req.body,
 *         signatureHeader: req.header('x-trove-signature')!,
 *       })
 *       ...
 *     })
 */
export function verifyWebhook<T = Record<string, unknown>>(
  opts: VerifyWebhookOptions,
): WebhookEvent<T> {
  const tolerance = opts.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS
  if (!opts.signatureHeader) {
    throw new WebhookSignatureError('Missing X-Trove-Signature header')
  }
  const { timestamp, v1 } = parseHeader(opts.signatureHeader)

  const now = Math.floor(Date.now() / 1000)
  if (Math.abs(now - timestamp) > tolerance) {
    throw new WebhookSignatureError(
      `Signature timestamp outside tolerance (${Math.abs(now - timestamp)}s > ${tolerance}s)`,
    )
  }

  const raw = typeof opts.body === 'string' ? Buffer.from(opts.body, 'utf8') : Buffer.from(opts.body)
  const prefix = Buffer.from(`${timestamp}.`, 'utf8')
  const payload = Buffer.concat([prefix, raw])

  const expected = createHmac('sha256', opts.secret).update(payload).digest('hex')
  const expectedBuf = Buffer.from(expected, 'hex')
  let providedBuf: Buffer
  try {
    providedBuf = Buffer.from(v1, 'hex')
  } catch {
    throw new WebhookSignatureError('Signature is not valid hex')
  }
  if (
    expectedBuf.length !== providedBuf.length ||
    !timingSafeEqual(expectedBuf, providedBuf)
  ) {
    throw new WebhookSignatureError('Signature mismatch')
  }

  let decoded: WebhookEvent<T>
  try {
    decoded = JSON.parse(raw.toString('utf8')) as WebhookEvent<T>
  } catch (e) {
    throw new WebhookSignatureError(`Body is not valid JSON: ${(e as Error).message}`)
  }
  return decoded
}
